### PR TITLE
[libcxx][docs] Make test name pattern documentation more obvious

### PR DESCRIPTION
**Pull request by hawkinsw**
As a new contributor, I found it hard to find the documentation for the meaning of the names of different tests and how those names translate to Lit.

This makes that documentation more explicit.

### CODE REVIEW

File: libcxx/docs/TestingLibcxx.rst
Comment (ldionne) [inline libcxx/docs/TestingLibcxx.rst:333]: I would like to avoid duplicating this documentation. It will get out of date when we add new test formats in the future.

I would suggest either migrating the whole documentation for our test format into `.rst` or dropping this change.

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/docs/TestingLibcxx.rst
Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst:333]: Okay!! Honestly, you tell me!! Which one would you prefer? I am happy to migrate it all and do a cleanup if you think it would be helpful!

Comment (ldionne) [reply libcxx/docs/TestingLibcxx.rst:333]: I think documenting the libc++ testing format in `.rst` would be more helpful to contributors than it staying hidden where it's currently at. I've had to show it to people several times in the past so I think this is a common one that people don't know about. I would still link to these docs from `format.py` though just to connect them in an obvious way.

**Comment by hawkinsw** [comment]
Great! Thank you for the "go ahead". I will send along some fixups and re-request review when it is ready!

**Comment by hawkinsw** [comment]
I have updated to consolidate all documentation about the file name pattern semantics to `docs/TestingLibcxx.rst`. In the future, I would like to expand the testing documentation to include information about how to write a *builtin Lit shell* test (including the valid commands, how to use pipelines, valid substitutions, etc). @ldionne : If you would prefer to wait for that additional documentation before merging this PR, that is great by me. If not, I will make a separate PR.

File: libcxx/docs/TestingLibcxx.rst
Comment (mordante) [inline libcxx/docs/TestingLibcxx.rst:338]: The last sentence seems no longer to be true.

File: libcxx/utils/libcxx/test/format.py
Comment (mordante) [inline libcxx/utils/libcxx/test/format.py:160]: I would only link to the website and use an anchor.

**Review by mordante — APPROVED** [review]
Thanks for working on this, I have grepped for this information a number times. Having it on the website gives it a better visibility.

LGTM modulo some comments. Please wait for @ldionne to review this too.

File: libcxx/docs/TestingLibcxx.rst
Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst:338]: Good point! :-)

**Comment by hawkinsw** [comment]
> Thanks for working on this, I have grepped for this information a number times. Having it on the website gives it a better visibility.
> 
> LGTM modulo some comments. Please wait for @ldionne to review this too.

Thank you for the feedback! I really appreciate it!

File: libcxx/utils/libcxx/test/format.py
Comment (hawkinsw) [reply libcxx/utils/libcxx/test/format.py:160]: Good point! Will do!

File: libcxx/docs/TestingLibcxx.rst
Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst]: Sorry! I had stupidly not committed a change that fixed this!! Should be resolved now!

File: libcxx/utils/libcxx/test/format.py
Comment (hawkinsw) [reply libcxx/utils/libcxx/test/format.py]: All done! Great feedback, thank you!

File: libcxx/utils/libcxx/test/format.py
Comment (ldionne) [inline libcxx/utils/libcxx/test/format.py:161]:
```suggestion
    Lit tests are contained in files that follow a certain pattern, which determines the semantics of the test.
    Under the hood, we basically generate a builtin Lit shell test that follows the ShTest format, and perform
    the appropriate operations (compile/link/run). See
    https://libcxx.llvm.org/TestingLibcxx.html#test-names
    for a complete description of those semantics.
```

File: libcxx/utils/libcxx/test/format.py
Comment (ldionne) [inline libcxx/utils/libcxx/test/format.py:163]: As a follow-up, it would be great to also move this part of the documentation to `.rst` so that all of the documentation for the file format is in the same place.

**Review by ldionne — APPROVED** [review]
LGTM w/ comments. It would be nice to move the rest of the documentation for the test format to `.rst` too, otherwise we'll have things in multiple places which means more likely to forget to update new information everywhere.

**Comment by hawkinsw** [comment]
> LGTM w/ comments. It would be nice to move the rest of the documentation for the test format to `.rst` too, otherwise we'll have things in multiple places which means more likely to forget to update new information everywhere.

Thank you for the review, @ldionne !! I will incorporate your feedback ASAP! The other thing that I would like to add with this PR is documentation on the actual commands available in the lit built-in shell tests. Does that seem reasonable? If so, I will start working on that and try to have something ready for tomorrow!

Thank you again!

**Comment by ldionne** [comment]
> The other thing that I would like to add with this PR is documentation on the actual commands available in the lit built-in shell tests.

Which commands do you mean? The actual builtin Lit ShTest commands like `RUN`? If so, those are already documented in Lit-specific documentation. We should move the documentation for the commands we add in our test format (like `ADDITIONAL_COMPILE_FLAGS`), though.

**Comment by hawkinsw** [comment]
> > The other thing that I would like to add with this PR is documentation on the actual commands available in the lit built-in shell tests.
> 
> Which commands do you mean? The actual builtin Lit ShTest commands like `RUN`? If so, those are already documented in Lit-specific documentation. We should move the documentation for the commands we add in our test format (like `ADDITIONAL_COMPILE_FLAGS`), though.

Great! I will work on that. 

I was talking about documenting the builtin lit "shell" commands (e.g., 

```Python
    inproc_builtins = {
        "cd": executeBuiltinCd,
        "export": executeBuiltinExport,
        "echo": executeBuiltinEcho,
        "@echo": executeBuiltinEcho,
        "mkdir": executeBuiltinMkdir,
        "popd": executeBuiltinPopd,
        "pushd": executeBuiltinPushd,
        "rm": executeBuiltinRm,
        ":": executeBuiltinColon,
    }   
```

and if/how they differ in their meaning from posix (or PowerShell). Those are not documented in the other lit documentation (e.g., https://llvm.org/docs/TestingGuide.html#writing-new-regression-tests) and it was something that I struggled to find. Would you prefer that I send that as PR directly to lit?

I will reflag you for a review when it is ready! Thanks again!

**Comment by hawkinsw** [comment]
> > The other thing that I would like to add with this PR is documentation on the actual commands available in the lit built-in shell tests.
> 
> Which commands do you mean? The actual builtin Lit ShTest commands like `RUN`? If so, those are already documented in Lit-specific documentation. We should move the documentation for the commands we add in our test format (like `ADDITIONAL_COMPILE_FLAGS`), though.

I have added documentation for the `ADDITIONAL_COMPILE_FLAGS` and `FILE_DEPENDENCIES` directives.

File: libcxx/docs/TestingLibcxx.rst
Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst]: Thank you! I have committed these!

File: libcxx/docs/TestingLibcxx.rst
Comment (ldionne) [inline libcxx/docs/TestingLibcxx.rst:406]: This new documentation is not accurate. They don't apply to Clang only, and they do affect `.sh.cpp` tests too. I suggest we stick to the previous documentation:

```
This directive will cause the provided flags to be added to the
%{compile_flags} substitution for the test that contains it. This
allows adding special compilation flags without having to use a
.sh.cpp test, which would be more powerful but perhaps overkill.
```

File: libcxx/docs/TestingLibcxx.rst
Comment (ldionne) [inline libcxx/docs/TestingLibcxx.rst:401]: This is slightly incorrect. This uses the `%T` substitution, not `%S`. I suggest keeping the wording in the old documentation, I think it wasn't unclear.

**Review by ldionne — CHANGES_REQUESTED** [review]

**Comment by hawkinsw** [comment]
@ldionne Thank you for the feedback. I am sorry that there were mistakes. I am *sure* that the mistakes were all my fault but I didn't intend to change any of the documentation that was previously written. I will address your comments as soon as possible. I am going to be away from my keyboard until Thursday but will work on it ASAP.

File: libcxx/docs/TestingLibcxx.rst
Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst:401]: The reason I wrote `%S` was 

```
# For each file dependency in FILE_DEPENDENCIES, inject a command to copy
# that file to the execution directory. Execute the copy from %S to allow
# relative paths from the test directory.
```

from line 130 of `format.py`. Sorry!

Comment (ldionne) [reply libcxx/docs/TestingLibcxx.rst:401]: Ah, I see -- you're right, this was missing from the old documentation. Maybe something like this?

```
This directive expresses that the test requires the provided files
or directories in order to run. An example is a test that requires
some test input stored in a data file. When a test file contains
such a directive, the files will collect them and copy them
to the directory represented by %T. The copy is performed from
the directory represented by %S (i.e. the source directory of  the
test being executed), which allows using relative paths to specify
files. After performing the copy, the intent is that %T contains all
the inputs necessary to run the test, such that e.g. execution on a
remote host can be done by simply copying %T to the host.
```

File: libcxx/utils/libcxx/test/format.py
Comment (hawkinsw) [reply libcxx/utils/libcxx/test/format.py:189]: All done!!

File: libcxx/docs/TestingLibcxx.rst
Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst:401]: Let me know what you think about what I am about to commit. I think that your suggestion was great -- the examples really help!

File: libcxx/docs/TestingLibcxx.rst
Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst:406]: I believe that I have properly incorporated this feedback as well! Thanks!

File: libcxx/docs/TestingLibcxx.rst
Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst:405]: I believe I got that in the latest fixup.

**Comment by hawkinsw** [comment]
@mordante In re: #76246 I have added some documentation for the `MODULE` directive. Let me know if it reads okay!

File: libcxx/docs/TestingLibcxx.rst
Comment (ldionne) [inline libcxx/docs/TestingLibcxx.rst:354]: I think you need to use double-backticks here for this to render properly!

**Review by ldionne — APPROVED** [review]
LGTM with minor changes to satisfy RST files. Thanks!

File: libcxx/docs/TestingLibcxx.rst
Comment (cjdb) [reply libcxx/docs/TestingLibcxx.rst:354]: I'm surprised this isn't a build error.

Comment (ldionne) [reply libcxx/docs/TestingLibcxx.rst:354]: I think it's valid RST code, it will just render weird.

Comment (hawkinsw) [reply libcxx/docs/TestingLibcxx.rst:354]: Of course I do!! Sorry. Thank you!

**Comment by hawkinsw** [comment]
Can I get one more check, @ldionne ? Thanks -- I'm just a little nervous!

**Review by ldionne — APPROVED** [review]
LGTM

**Pull request merged**
